src/data-table.js/data-table.js

```diff
@@ -0,0 +1,338 @@
+  const [selection, setSelection] = React.useState<Set<string>>(new Set());
+  const [exclude, setExclude] = React.useState(false);
+
+  const categories = React.useMemo(() => {
+    return props.data.reduce((set, category) => set.add(category), new Set());
+  }, [props.data]);
+
+  return (
+    <div>
+      {Array.from(categories, (category, i) => (
+        <Checkbox
+          key={i}
+          checked={selection.has(category)}
+          onChange={() => {
+            if (selection.has(category)) {
+              selection.delete(category);
+            } else {
+              selection.add(category);
+            }
+            setSelection(new Set(selection));
+          }}
+        >
+          {category}
+        </Checkbox>
+      ))}
+      <div className={useCss({display: 'flex'})}>
+        <Checkbox
+          checked={exclude}
+          onChange={() => setExclude(!exclude)}
+          checkmarkType={STYLE_TYPE.toggle}
+        >
+          Exclude
+        </Checkbox>
+        <Button
+          onClick={() => {
+            props.setFilter(
+              {selection, exclude},
+              Array.from(selection).join(', '),
+            );
+            props.close();
+          }}
+        >
+          Apply
+        </Button>
+      </div>
+    </div>
+  );
+}
+
+export function DataTable(props: Props) {
+  const [sortIndex, setSortIndex] = React.useState(-1);
+  const [sortDirection, setSortDirection] = React.useState(null);
+  const [filters, setFilters] = React.useState(new Map());
+
+  React.useEffect(() => {
+    const titles = props.columns.reduce(
+      (set, column) => set.add(column.title),
+      new Set(),
+    );
+    if (titles.size < props.columns.length) {
+      console.warn(
+        'Columns titles must be unique else will result in non-deterministic filtering.',
+      );
+    }
+  }, [props.columns]);
+
+  function handleSort(columnIndex) {
+    if (columnIndex === sortIndex) {
+      if (sortDirection === 'ASC') {
+        setSortIndex(-1);
+        setSortDirection('DESC');
+      } else {
+        setSortDirection('ASC');
+      }
+    } else {
+      setSortIndex(columnIndex);
+      setSortDirection('DESC');
+    }
+  }
+
+  function addFilter(filterParams, title, description) {
+    filters.set(title, {filterParams, description});
+    setFilters(new Map(filters));
+  }
+
+  function removeFilter(title) {
+    filters.delete(title);
+    setFilters(new Map(filters));
+  }
+
+  const rows = React.useMemo(() => {
+    if (sortIndex === -1 && !filters.size) {
+      return props.rows;
+    }
+
+    let nextRows = [].concat(props.rows);
+    if (sortIndex !== -1) {
+      const sortFn = sortFnByColumn(props.columns[sortIndex]);
+      nextRows.sort((a, b) => sortFn(a.data[sortIndex], b.data[sortIndex]));
+
+      if (sortDirection === 'ASC') {
+        nextRows = nextRows.reverse();
+      }
+    }
+
+    nextRows = Array.from(filters, f => f).reduce((rows, [title, filter]) => {
```

Comment:
some shower thoughts:
it may be more performant to separate out filter/sort transformations. where sorting can be modeled as an array of row indices and filtering as a set of row indices. what's nice is that changing a filter/sort only affects it's corresponding array/set, rather than needing to perform both transforms when either change.